src/parcsr_ls/par_cheby.c

```diff
@@ -38,130 +38,136 @@ means half, and .1 means 10percent)
 
 *******************************************************************************/
 
-HYPRE_Int hypre_ParCSRRelax_Cheby_Setup(hypre_ParCSRMatrix *A, /* matrix to relax with */
-                            HYPRE_Real max_eig,      
-                            HYPRE_Real min_eig,     
-                            HYPRE_Real fraction,   
-                            HYPRE_Int order,            /* polynomial order */
-                            HYPRE_Int scale,            /* scale by diagonal?*/
-                            HYPRE_Int variant,           
-                            HYPRE_Real **coefs_ptr,
-                            HYPRE_Real **ds_ptr)   /* initial/updated approximation */
+HYPRE_Int
+hypre_ParCSRRelax_Cheby_Setup(hypre_ParCSRMatrix *A,         /* matrix to relax with */
+                              HYPRE_Real          max_eig,
+                              HYPRE_Real          min_eig,
+                              HYPRE_Real          fraction,
+                              HYPRE_Int           order,     /* polynomial order */
+                              HYPRE_Int           scale,     /* scale by diagonal?*/
+                              HYPRE_Int           variant,
+                              HYPRE_Real        **coefs_ptr,
+                              HYPRE_Real        **ds_ptr)    /* initial/updated approximation */
 {
-   hypre_CSRMatrix *A_diag = hypre_ParCSRMatrixDiag(A);
+   hypre_CSRMatrix *A_diag       = hypre_ParCSRMatrixDiag(A);
    HYPRE_Real      *A_diag_data  = hypre_CSRMatrixData(A_diag);
    HYPRE_Int       *A_diag_i     = hypre_CSRMatrixI(A_diag);
-
-   HYPRE_Real theta, delta;
-   
-   HYPRE_Real den;
-   HYPRE_Real upper_bound, lower_bound;
-   
-   HYPRE_Int j;
-   HYPRE_Int num_rows = hypre_CSRMatrixNumRows(A_diag);
- 
-   HYPRE_Real *coefs = NULL;
-   
-   HYPRE_Int cheby_order;
-
-   HYPRE_Real *ds_data = NULL;
-   HYPRE_Real  diag;
+   HYPRE_Real       theta, delta;
+   HYPRE_Real       den;
+   HYPRE_Real       upper_bound = 0.0, lower_bound = 0.0;
+   HYPRE_Int        j;
+   HYPRE_Int        num_rows     = hypre_CSRMatrixNumRows(A_diag);
+   HYPRE_Real      *coefs        = NULL;
+   HYPRE_Int        cheby_order;
+   HYPRE_Real      *ds_data = NULL;
+   HYPRE_Real       diag;
 
    /* u = u + p(A)r */
-
    if (order > 4)
+   {
       order = 4;
+   }
+
    if (order < 1)
+   {
       order = 1;
+   }
 
-   coefs = hypre_CTAlloc(HYPRE_Real,  order+1, HYPRE_MEMORY_HOST);
+   coefs = hypre_CTAlloc(HYPRE_Real, order+1, HYPRE_MEMORY_HOST);
    /* we are using the order of p(A) */
-   cheby_order = order -1;
-   
-    /* make sure we are large enough -  Adams et al. 2003 */
-   upper_bound = max_eig * 1.1;
-   /* lower_bound = max_eig/fraction; */
-   lower_bound = (upper_bound - min_eig)* fraction + min_eig; 
+   cheby_order = order - 1;
 
+   if (min_eig >= 0.0)
```

Comment:
I don't think this is quite right.  Maybe we should discuss.  Basically, what was identified as `upper_bound` in the previous code was actually the eigenvalue estimate (positive or negative) that corresponds to the eigenvalue with largest magnitude.  I believe the Chebyshev polynomial computations were correct in the previous code even in the negative definite case.  I don't think they are anymore here.  One other thing to fix is "`min_eig`", which I think we are better off setting to zero as in the previous code.

If you want to think about this ahead of time, just consider the simplest zero-th order case that basically produces the Richardson error propagator (referred to as the residual polynomial in the comments below), I - sA, where s is the inverse of the eigenvalue (positive or negative) with largest magnitude.  See `theta` and the resulting `coefs[0]` computation; the latter is s for Richardson.  This provides a good smoother when the problem is either positive or negative (semi-)definite.  I don't know what to do when the problem is indefinite.  Smoothers in that setting or more difficult and usually involve the normal equations.

Reply:
I see that you are already setting one of the estimates to zero, so ignore that comment above.